core/src/main/java/com/segment/analytics/kotlin/core/platform/plugins/logger/LogTarget.kt

```diff
@@ -0,0 +1,238 @@
+        }
+    }
+}
+
+/**
+ * The Segment logging system has three types of logs: log, metric and history. When adding a target that
+ * responds to logs, it is possible to adhere to 1 to many. In other words, a LoggingType can be .log &
+ * .history. This is used to tell which targets logs are directed to.
+ */
+class LoggingType(types: List<Filter>) {
+
+    enum class Filter {
+        LOG,
+        METRIC,
+        HISTORY;
+    }
+
+    companion object {
+        /// Convenience .log logging type
+        val log = LoggingType(listOf(Filter.LOG))
+        /// Convenience .metric logging type
+        val metric = LoggingType(listOf(Filter.METRIC))
+        /// Convenience .history logging type
+        val history = LoggingType(listOf(Filter.HISTORY))
+    }
+
+    // - Private Properties and Methods
+    private val allTypes: List<Filter> = types
+
+    /**
+     * Convenience method to find if the LoggingType supports a particular destination.
+     *
+     * @property destination The particular destination being tested for conformance.
+     */
+    internal fun contains(destination: Filter): Boolean {
+        return this.allTypes.contains(destination)
+    }
+}
+
+/**
+ * The interface to the message being returned to `LogTarget` -> `parseLog()`.
+ */
+interface LogMessage {
+    val kind: LogFilterKind
+    val title: String?
+    val message: String
+    val event: BaseEvent?
+    val function: String?
+    val line: Int?
+    val logType: LoggingType.Filter
+    val dateTime: Date
+}
+
+enum class MetricType(val type: Int) {
+    Counter(0), // Not Verbose
+    Gauge(1);   // Semi-verbose
+
+    override fun toString(): String {
+        var typeString = "Gauge"
+        if (this == Counter) {
+            typeString = "Counter"
+        }
+        return typeString
+    }
+
+    companion object {
+        fun from(string: String): MetricType {
+            var returnType = Counter
+            if (string == "Gauge") {
+                returnType = Gauge
+            }
+            return returnType
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/**
+ * The public logging method for capturing all general types of log messages related to Segment.
+ *
+ * @property  message The main message of the log to be captured.
+ * @property kind Usually .error, .warning or .debug, in order of severity. This helps filter logs based on
+ * this added metadata.
+ * @property function The name of the function the log came from. This will be captured automatically.
+ * @property line The line number in the function the log came from. This will be captured automatically.
+ */
+@JvmOverloads
+fun Analytics.log(message: String, kind: LogFilterKind? = null, function: String = "", line: Int = -1) {
+    applyClosureToPlugins { plugin: Plugin ->
+        // Check if we should send the event
+        if (!SegmentLog.loggingEnabled) {
```

Comment:
it's probably better to do this check before line 131 `applyClosureToPlugins`, so that we have an earlier termination.  otherwise, it goes through unnecessary loops that iterates the plugins.

Reply:
Yep good catch